File: 3_Add two numbers.py
# 2 . Add two numbers
# You are given two non-empty linked lists representing two non-negative integers. 
# The digits are stored in reverse order, and each of their nodes contains a single digit. 
# Add the two numbers and return the sum as a linked list.
# You may assume the two numbers do not contain any leading zero, except the number 0 itself.


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        output = ListNode()
        output1 = output
        while l1 or l2 or carry:
            val1 = (l1.val if l1 else 0) 
            val2 = (l2.val if l2 else 0)
            
            sum = val1 + val2 + carry
            val = sum % 10
            carry = sum // 10
            output1.next = ListNode(val)
            output1 = output1.next
            
            l1 = (l1.next if l1 else None)
            l2 = (l2.next if l2 else None)
        return output.next

File: 3_Add two numbers.java
// # 2 . Add two numbers
// # You are given two non-empty linked lists representing two non-negative integers. 
// # The digits are stored in reverse order, and each of their nodes contains a single digit. 
// # Add the two numbers and return the sum as a linked list.
// # You may assume the two numbers do not contain any leading zero, except the number 0 itself.


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode first = l1, second = l2;
        int carry = 0;
        ListNode dummy = new ListNode(0);
        ListNode trav = dummy;
        while(first != null || second != null || carry!= 0){
            if (first != null){
                carry += first.val;
                first = first.next;
            }
            if (second != null){
                carry +=second.val;
                second = second.next;
            }
            
            trav.next = new ListNode(carry % 10);
            trav = trav.next;
            carry /=10;
            }
        
        return dummy.next;
    }
}